rfcs/2021-19-02-units-for-typography-and-layout.md

```diff
@@ -0,0 +1,603 @@
+<br/> <br/>
+
+> 🔗 You can play around with all the [demos here](https://szi8i.csb.app/)
+
+#### For React Native <!-- omit in toc -->
+It's **layout in dp and typography in dp w/ autoscale**. Why?
+* React native platform only allows `dp` as units.
+* It's better since the layouts become much more predictable as the platform just asks us to define things at a base layout and takes care of scaling on different devices with different screen densities automatically.
+* We will always honor user's device font size since we want the content to be readable as per user's preference. It might break the layout in some cases but the cons weighs off the pros i.e **content readability**
+
+Let's tally our decision with our [checklist](#checklist)
+1. Whether the content is accessible when the user changes the default font size of the browser/device ✅
+2. Whether the content is accessible when the user zooms in/out on the browser ✅
+3. Whether the interface is able to render on different screen sizes, pixel densities etc ✅
+
+
+## How will we store and use it?
+<img alt="storing and getting tokens" src="./images/unit-storing-and-getting.png" width="1200px">
+
+* While storing we can compute the value in `px/dp` and store it in a unitless way. While rendering we can attach the units.
+   ```js
+   const space = {
+      1: 8,
+      2: 12,
+      3: 16,
+   }
+   const fontSize = {
+      1: 14,
+      2: 18,
+      3: 24,
+   }
+   ```
+* With the above approach we can keep the vocabulary consistent that everything is stored in `px` for web and `dp` for react native  but while rendering the respective platforms(web/react-native) attaches the unit for the target rendering engine.
+* We will implement a generic funtion that will attach the units to the tokens.
+* For typography the units that'll be constructed will be relative(`rems` for web and `autoScale` value on text enabled for react-native apps).
+* For layout i.e height, width, padding, margin the units that'll be constructed will be absolute i.e `px`(pixels) for web and `dp` for react-native apps but with device pixel ratio to ensure consistency with various screen densities.
+# Drawbacks/Constraints
+- Vocabulary is the biggest drawback. Thinking and visualising in relative units is difficult compared to absolute units like pixels(`px`) or density independent pixels(`dp`) but we can work it out with the help of tools. We can think and store in pixels but render in relative units of the target platform(`rems` for web and `dp` with device pixel ratio for react-native apps).
+
+# Alternatives
+- Here's the [absolute unit vs relative unit matrix](#absolute-unit-vs-relative-unit-matrix) which states all the altternatives with the details.
+
+# Adoption strategy
+- **For developers** - All these decisions shouldn't matter, as this will be stored in our tokens as an abstraction. The end consumers will just use the typography and layout scale tokens for their purpose and not the actual values.
```

Comment:
How will people figure out which value to use from the scale (without manual calculations in head or checking against the token values)? Say if Figma shows, 32px spacing in between two cards. Would we be able to solve this with some Figma configs / plugins, etc.?

Reply:
So once the Figma is all set people will see the actual token scale name in the inspector on Figma. So that'll help us to avoid all these confusion as devs will just care to look at what scale value they need to pick instead of the actual px/dp value